core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -0,0 +1,381 @@
+    mode = nextLockMode(mode);
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(!endsInInode());
+      String lastEdgeName = ((EdgeEntry) lastEntry()).getEdge().getName();
+      Preconditions.checkState(inode.getName().equals(lastEdgeName),
+          "Expected to lock inode %s but locked inode %s", lastEdgeName, inode.getName());
+    }
+    addEntry(createInodeEntry(inode, mode));
+  }
+
+  @Override
+  public void lockEdge(Inode lastInode, String childName, LockMode mode) {
+    mode = nextLockMode(mode);
+    long edgeParentId = lastInode.getId();
+    if (!mEntries.isEmpty()) {
+      Preconditions.checkState(endsInInode());
+      Preconditions.checkState(lastInode().getId() == edgeParentId);
+    }
+    addEntry(createEdgeEntry(new Edge(edgeParentId, childName), mode));
+  }
+
+  @Override
+  public void lockRootEdge(LockMode mode) {
+    Preconditions.checkState(mEntries.isEmpty());
+    addEntry(createEdgeEntry(ROOT_EDGE, mode));
+    if (mode == LockMode.WRITE) {
+      mFirstWriteLockIndex = 0;
+    }
+  }
+
+  @Override
+  public void pushWriteLockedEdge(Inode inode, String childName) {
+    Preconditions.checkState(!endsInInode());
+    Preconditions.checkState(endsInWriteLock());
+
+    if (endsInMultipleWriteLocks()) {
+      // If the lock before the edge lock is already WRITE, we can just acquire more WRITE locks.
+      lockInode(inode, LockMode.WRITE);
+      lockEdge(inode, childName, LockMode.WRITE);
+    } else {
+      EdgeEntry lastEdgeEntry = createEdgeEntry(lastEdge(), LockMode.READ);
+      InodeEntry inodeEntry = createInodeEntry(inode, LockMode.READ);
+      EdgeEntry nextEdgeEntry = createEdgeEntry(new Edge(inode.getId(), childName), LockMode.WRITE);
+      removeLastEntry(); // Remove edge write lock
+      addEntry(lastEdgeEntry);
+      addEntry(inodeEntry);
+      addEntry(nextEdgeEntry);
+    }
+  }
+
+  @Override
+  public void unlockLastInode() {
+    Preconditions.checkState(endsInInode());
+    Preconditions.checkState(!mEntries.isEmpty());
+    removeLastEntry();
+  }
+
+  @Override
+  public void unlockLastEdge() {
+    Preconditions.checkState(!endsInInode());
+    Preconditions.checkState(!mEntries.isEmpty());
+    removeLastEntry();
+  }
+
+  @Override
+  public void downgradeLastInode() {
+    Preconditions.checkState(endsInInode());
+    Preconditions.checkState(!mEntries.isEmpty());
+    Preconditions.checkState(endsInWriteLock());
+
+    if (endsInMultipleWriteLocks()) {
+      return; // No point in changing from write lock to read lock when the parent lock is WRITE.
+    }
+
+    InodeEntry newEntry = createInodeEntry(lastInode(), LockMode.READ);
+    removeLastEntry();
+    addEntry(newEntry);
+  }
+
+  @Override
+  public void downgradeLastEdge() {
+    Preconditions.checkState(!endsInInode());
+    Preconditions.checkState(!mEntries.isEmpty());
+    Preconditions.checkState(endsInWriteLock());
+
+    if (!endsInMultipleWriteLocks()) {
```

Comment:
NIT: can we mirror the logic in `downgradeLastInode`?

Reply:
done